Source/JavaScriptCore/builtins/ArrayConstructor.js

```diff
@@ -110,3 +111,95 @@ function isArray(array)
         return false;
     return @isArraySlow(array);
 }
+
+@linkTimeConstant
+async function defaultAsyncFromAsyncIterator(result, iterator, mapFn, thisArg)
+{
+    "use strict";
+
+    var k = 0;
+
+    // Since for-of loop once more looks up the @@iterator property of a given iterable,
+    // it could be observable if the user defines a getter for @@iterator.
+    // To avoid this situation, we define a wrapper object that @@iterator just returns a given iterator.
+    var wrapper = {
+        @@asyncIterator: function () { return iterator; }
+    };
+
+    for await (var value of wrapper) {
+        if (k >= @MAX_SAFE_INTEGER)
+            @throwTypeError("Length exceeded the maximum array length");
+        if (mapFn)
+            @putByValDirect(result, k, await (thisArg === @undefined ? mapFn(value, k) : mapFn.@call(thisArg, value, k)));
+        else
+            @putByValDirect(result, k, value);
+        k += 1;
+    }
+
+    result.length = k;
+    return result;
+}
+
+@linkTimeConstant
+async function defaultAsyncFromAsyncArrayLike(asyncItems, mapFn, thisArg)
+{
+    "use strict";
+
+    var arrayLike = @toObject(asyncItems, "Array.fromAsync requires an array-like object - not null or undefined");
+
+    var arrayLikeLength = @toLength(arrayLike.length);
+
+    var result = this !== @Array && @isConstructor(this) ? new this(arrayLikeLength) : @newArrayWithSize(arrayLikeLength);
+
+    var k = 0;
+    while (k < arrayLikeLength) {
+        var value = await arrayLike[k];
+        if (mapFn)
+            @putByValDirect(result, k, await (thisArg === @undefined ? mapFn(value, k) : mapFn.@call(thisArg, value, k)));
+        else
+            @putByValDirect(result, k, value);
+        k += 1;
+    }
+
+    result.length = arrayLikeLength;
+    return result;
+}
+
+function fromAsync(asyncItems  /*, mapFn, thisArg */)
+{
+    "use strict";
+
+    try {
+        var mapFn = @argument(1);
+
+        var thisArg;
+
+        if (mapFn !== @undefined) {
+            if (!@isCallable(mapFn))
+                @throwTypeError("Array.from requires that the second argument, when provided, be a function");
+
+            thisArg = @argument(2);
+        }
+
+        var usingAsyncIterator = asyncItems.@@asyncIterator;
```

Comment:
`GetMethod` throw `TypeError` for non-callable's and also coerces `null` to `undefined`.